### PR TITLE
[luci] Introduce RemoveUnnecessaryReshapeNetPass

### PR DESCRIPTION
**Pull request by jinevening**
This introduces a RemoveUnnecessaryReshapeNet Pass.

ONE-DCO-1.0-Signed-off-by: Hyukjin Jeong <hj1.jeong@samsung.com>

---
Related to: https://github.com/Samsung/ONE/issues/9600

### CODE REVIEW

File: compiler/luci/pass/src/RemoveUnnecessaryReshapeNetPass.cpp [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    assert(post_reshape != nullptr); // FIX_CALLER_UNLESS
+  }
+
+public:
+  // Return true if pre_reshapes are found
+  bool collect_pre_reshapes(loco::Node *node)
+  {
+    // TODO Support diamond case
+    if (loco::succs(node).size() != 1)
+      return false;
```
Comment (jinevening) [inline compiler/luci/pass/src/RemoveUnnecessaryReshapeNetPass.cpp:76]: Condition 4 in https://github.com/Samsung/ONE/issues/9600#issuecomment-1220258297

File: compiler/luci/pass/src/RemoveUnnecessaryReshapeNetPass.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+    if (auto pre_reshape = dynamic_cast<luci::CircleReshape *>(node))
+    {
+      // Check ifm of pre-reshape and ofm of post_reshape
+      if (not same_shape(pre_reshape->tensor(), _post_reshape))
+        return false;
+
+      // Check ofm of pre-reshape and ifm of post_reshape
+      if (not same_shape(pre_reshape, _post_reshape->tensor()))
+        return false;
```
Comment (jinevening) [inline compiler/luci/pass/src/RemoveUnnecessaryReshapeNetPass.cpp:86]: Condition 1, 2 in https://github.com/Samsung/ONE/issues/9600#issuecomment-1220258297

File: compiler/luci/pass/src/RemoveUnnecessaryReshapeNetPass.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+      _pre_reshapes.emplace_back(pre_reshape);
+      return true;
+    }
+
+    if (not acceptable_intermediate_op(node))
```
Comment (jinevening) [inline compiler/luci/pass/src/RemoveUnnecessaryReshapeNetPass.cpp:92]: Condition 3 in https://github.com/Samsung/ONE/issues/9600#issuecomment-1220258297

**Review by seanshpark — APPROVED** [review]
LGTM thank you!

**Pull request merged**